### PR TITLE
docs: Add Local Inference findings

### PR DESCRIPTION
**Pull request by tchetwin**

### CODE REVIEW

**Review by robpalme — APPROVED** [review]

File: external-declarations/docs/local-inferences.md [hunk 1]
```diff
@@ -0,0 +1,1063 @@
...
+- **User complexity if not provided:** Low
+- **Emitter complexity:** Low
+
+### INF10: JSX elements ***(TODO: Still under investigation)***
+
+We can’t fully emulate the JSX namespace resolution TypeScript does, but we can make a reasonable guess.
+
+<table><th>Source</th><th>Inferred declaration</th><tbody><td>
+
+```ts
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
```tsx
```

File: external-declarations/docs/local-inferences.md [hunk 2]
```diff
@@ -0,0 +1,1063 @@
+# Local Inferences for `isolatedDeclarations`
+
+An initial implementation of `isolatedDeclarations` did not perform any type of inference.
+This _"pure approach"_ was simple to implement but was not ergonomic due to the number of explicit type annotations it demanded.
+To improve the developer experience, inferences were added and are described in this document.
+
+These inferences are implemented in [the reference ***Declaration Type Emitter (DTE)***](../src/compiler) and are verified by [the conformance tests](./tests).
```
Comment (frigus02) [inline external-declarations/docs/local-inferences.md]:
```suggestion
These inferences are implemented in [the reference ***Declaration Type Emitter (DTE)***](../src/compiler) and are verified by [the conformance tests](../tests).
```

File: external-declarations/docs/local-inferences.md [hunk 3]
```diff
@@ -0,0 +1,1063 @@
...
+There is no cross-file analysis (this keeps declaration emit parallelizable) and no in-file analysis beyond the current expression.
+While some in-file analysis might help, it would require replicating a lot of TypeScript machinery and so is avoided.
+
+### P2. Optimistic emit
+
+Sometimes a DTE may not have enough information to determine if the naturally emitted declaration is correct.
+In such cases declaration emit can proceed with a deterministic inference and rely on the user to run the type-checker to detect the error.
+Meaning the emitter may complete its output but we have to wait for the type-checker to be sure the declarations are valid.
+
+Optimistic emit should only happen in cases where we deem that the likelihood is high that the type is correct.Some examples:
```
Comment (frigus02) [inline external-declarations/docs/local-inferences.md]:
```suggestion
Optimistic emit should only happen in cases where we deem that the likelihood is high that the type is correct. Some examples:
```

File: external-declarations/docs/local-inferences.md [hunk 4]
```diff
@@ -0,0 +1,1063 @@
...
+type Id = { id: string, foo?: string }
+
+// TS: let a: Id[]
+let a = [
+    { id: "" } as Id,
+    { id:"", foo: "" },
+]
+```
+
+Typescript can reduce the type of the array element from `Id | { id: string, foo: string }` to just `Id`.
```
Comment (frigus02) [inline external-declarations/docs/local-inferences.md]:
```suggestion
TypeScript can reduce the type of the array element from `Id | { id: string, foo: string }` to just `Id`.
```

**Review by frigus02 — APPROVED** [review]
Great work!

File: external-declarations/docs/local-inferences.md [hunk 5]
```diff
@@ -0,0 +1,1063 @@
...
+  bar: {
+    baz: 1
+  }
+}, {
+  bar: {
+    bat: 1
+  }
+}];
+
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
 
```

File: external-declarations/docs/local-inferences.md [hunk 6]
```diff
@@ -0,0 +1,1063 @@
...
+  foo: string;
+  bar: number;
+} | {
+  foo: string;
+  bar?: undefined;
+};
+
+
+
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
 
 
 
```

File: external-declarations/docs/local-inferences.md [hunk 7]
```diff
@@ -0,0 +1,1063 @@
...
+<table><th>Source</th><th>Inferred declaration</th><tbody><td>
+
+```ts
+const LogLevel = {
+  ...HappyLogLevel,
+  ...UnhappyLogLevel,
+  UNDEFINED: "UNDEFINED",
+} as const;
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
```

File: external-declarations/docs/local-inferences.md [hunk 8]
```diff
@@ -0,0 +1,1063 @@
...
+```ts
+let tuple = [1, 2, 3] as const
+const composedTuple = [0, ...tuple] as const
+```
+
+</td><td>
+
+```ts
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
```

File: external-declarations/docs/local-inferences.md [hunk 9]
```diff
@@ -0,0 +1,1063 @@
...
+We can infer the types added on to function variables using the same inference we use for other initializers.
+This is optimistic if any inferred property types are optimistic.
+
+<table><th>Source</th><th>Inferred declaration</th><tbody><td>
+
+```ts
+export const x = () => {}
+x.test = 1;
+
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
 
```

File: external-declarations/docs/local-inferences.md [hunk 10]
```diff
@@ -0,0 +1,1063 @@
...
+</td><td>
+
+```ts
+declare let service: {
+  request(): number;
+};
+
+
+
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
 
 
 
```

File: external-declarations/docs/local-inferences.md [hunk 11]
```diff
@@ -0,0 +1,1063 @@
...
+}
+```
+
+</td><td>
+
+```ts
+declare function test(p: number): () => number
+
+
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
 
 
```

**Review by robpalme — COMMENTED** [review]
Restore whitespace padding for alignment.

File: external-declarations/docs/local-inferences.md [hunk 12]
```diff
@@ -0,0 +1,1063 @@
...
+}
+```
+
+</td><td>
+
+```ts
+let fn: (a: number) => void
+let fn2: (b: string) => string
+
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md]:
```suggestion
 
 
```

File: external-declarations/docs/local-inferences.md [hunk 13]
```diff
@@ -0,0 +1,1063 @@
...
+```
+
+</td></tbody></table>
+
+- **Optimistic:** No
+- **User complexity if not provided:** Low
+- **Emitter complexity:** Low
+
+### INF04: Symbols (widening and non widening)
+
```
Comment (robpalme) [inline external-declarations/docs/local-inferences.md:102]:
```suggestion

Note: We need to check whether [TypeScript's existing widening for unique symbol](https://www.typescriptlang.org/play?target=99&ts=5.2.0-dev.20230523#code/PTAEAEBcEMCcHMCmkBcpEGcB2iAekAoECAE0QGMAbOaSASwHssC8AHB2SUcpjLjAJ4BbUAF5QAZWEAjBpQAUASgDcLXO07deXBtIBWY0AG8CoM6GhpBQggF8LGLVj6q2HLj2f9hAGTpYAa0NdPQA6aFUgA) is intentional.

```

**Pull request merged**